File: hosts/python/src/teml_host/paint.py
"""ONLCR-safe terminal painting."""

from __future__ import annotations

import sys
from typing import TextIO

from teml_host.screen import ScreenBuffer

_CLEAR_ALL = "\x1b[2J\x1b[H"


def paint(screen: ScreenBuffer, stream: TextIO | None = None) -> None:
    out = stream or sys.stdout
    out.write(_CLEAR_ALL)
    out.write(screen.text().replace("\n", "\r\n"))
    out.flush()

File: crates/teml-host/src/paint.rs
//! ONLCR-safe full-screen repaint helpers.

use crate::screen::ScreenBuffer;
use std::io::{self, Write};

/// Expand `\n` to `\r\n` so raw-mode hosts do not staircase across the terminal.
pub fn onlcr(text: &str) -> String {
    text.replace('\n', "\r\n")
}

/// Clear the screen, move to the origin, and write ONLCR-safe frame text.
pub fn paint<W: Write>(screen: &ScreenBuffer, mut out: W) -> io::Result<()> {
    #[cfg(feature = "terminal")]
    {
        use crossterm::{
            cursor::MoveTo,
            terminal::{Clear, ClearType},
            ExecutableCommand,
        };
        out.execute(Clear(ClearType::All))?.execute(MoveTo(0, 0))?;
    }
    write!(out, "{}", onlcr(&screen.text()))?;
    out.flush()
}
